lucene/core/src/java/org/apache/lucene/search/QueryVisitor.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import org.apache.lucene.index.Term;
+
+/**
+ * Interface to allow recursion through a query tree
+ *
+ * @see Query#visit(QueryVisitor)
+ */
+public interface QueryVisitor {
+
+  /**
+   * Called by leaf queries that match on a specific term
+   *
+   * @param term  the term the query will match on
+   */
+  void matchesTerm(Term term);
+
+  /**
+   * Called by leaf queries that do not match on terms
+   * @param query the query
+   */
+  default void visitLeaf(Query query) {}
+
+  /**
+   * Pulls a visitor instance for visiting matching child clauses of a query
+   *
+   * The default implementation returns {@code this}
+   *
+   * @param parent  the query visited
+   */
+  default QueryVisitor getMatchingVisitor(Query parent) {
```

Comment:
this corresponds to `BooleanClause.Occur` maybe we should just have a method `default void visitLeaf(Query query, BooleanClause.Occur occur) {}` instead of 4 default methods?

Reply:
oh yeah and a `matchesTerm(Term, Occur)` as well.

Reply:
now that I looked further down I think I see what you are trying to do here. I wonder if it makes sense to reduce to a single `getSubVisitor(BooleanClause.Occur occur, Query parent)` this seems more straight forward?

Reply:
++, have pushed a change to implement this